azure/azure_kv.go

```diff
@@ -98,8 +113,13 @@ func (az *azureSecrets) GetSecret(
 	}
 
 	t := func() (interface{}, bool, error) {
-		secretResp, err := az.kv.GetSecret(ctx, az.baseURL, secretID, "")
+		// passing empty version to always get the latest version of the secret.
+		secretResp, err := az.kv.GetSecret(ctx, secretID, "", nil)
 		if err != nil {
+			// don't retry if Secret is not present
+			if strings.Contains(err.Error(), "SecretNotFound") {
+				return nil, false, secrets.ErrSecretNotFound
+			}
```

Comment:
It can be fragile to rely on errors from libraries to contain specific strings. It would be best to modify the library to return an error of type `SecretNotFoundError` or something similar with its own internal logic for determining if the error is the same. 

The implementation should be resilient to modifications from localization frameworks that change text strings to output information in different languages.

Reply:
Use case: 
- Before we add the secret to the Azure KMS, we need to ensure that the secret is not already present. Else, it will keep adding versions of the secret. To ensure this, we first make a `get` request. If the Secret is not found only then we go ahead add the secret. 

Azure KMS support in this library uses a retry with timeout mechanism.  I feel we won't need that if the secret key was not found  (that is `404 Not found` or `SecretNotFound`).  Apart from this, we also need a way for this library to return `ErrSecretNotFound` so that users of this lib know that secret is not found.

Reply:
Agree that this is not the best way to do it.  The official Azure library is returning the `httpResponse` as error if the `http status` is not `OK`.   So this error returned by Azure, contains the http code of 404, `SecretNotFound` and a help message. 

I will check if there is a better way to handle this.

Reply:
Kubernetes' `IsAlreadyExists()` error checker is a good reference, plus other things in the same file. IIRC, k8s uses status codes (404 here) as an underlying mechanism for detecting which error is happening.

Reply:
The string format of the error response that we currently get from the azure sdk is like this: 

``` 
RESPONSE 404: 404 Not Found
ERROR CODE: SecretNotFound
--------------------------------------------------------------------------------
{
  "error": {
    "code": "SecretNotFound",
    "message": "A secret with (name/id) mysecret12 was not found in this key vault. If you recently deleted this secret you may be able to recover it using the correct recovery command. For help resolving this issue, please see https://go.microsoft.com/fwlink/?linkid=2125182"
  }
}
``` 

They are implementing the `Error()` method of the `error` interface and converting the httpResponse to `error` 


I'll take a look at the `IsAlreadyExists()` method.

Reply:
marking this as resolved. Using the `statuscode` to verify if secret is found or not.